api/liboni/drivers/oepcie/liboni-driver-oepcie.c

```diff
@@ -0,0 +1,362 @@
+	if (ctx->write.fid == -1) {
+		//fprintf(stderr, "%s: %s\n", strerror(errno), ctx->write.path);
+		close(ctx->config.fid);
+		close(ctx->signal.fid);
+		close(ctx->read.fid);
+		return ONI_EPATHINVALID;
+	}
+	ctx->file_state = OPEN;
+
+	return ONI_ESUCCESS;
+}
+
+int oni_driver_destroy_ctx(oni_driver_ctx driver_ctx)
+{
+	CTX_CAST;
+
+	if (ctx->file_state >= OPEN) {
+
+		if (close(ctx->config.fid) == -1) goto oni_close_ctx_fail;
+		if (close(ctx->read.fid) == -1) goto oni_close_ctx_fail;
+		if (close(ctx->write.fid) == -1) goto oni_close_ctx_fail;
+		if (close(ctx->signal.fid) == -1) goto oni_close_ctx_fail;
+	}
+
+	free(ctx->config.path);
+	free(ctx->read.path);
+	free(ctx->write.path);
+	free(ctx->signal.path);
+	free(ctx);
+
+	return ONI_ESUCCESS;
+
+oni_close_ctx_fail:
+	return ONI_ECLOSEFAIL;
+}
+
+int oni_driver_read_stream(oni_driver_ctx driver_ctx, oni_read_stream_t stream, void* data, size_t size)
+{
+	CTX_CAST;
+	size_t received = 0;
+
+	int data_fd;
+	switch (stream)
+	{
+	case ONI_READ_STREAM_DATA:
+		data_fd = ctx->read.fid;
+		break;
+	case ONI_READ_STREAM_SIGNAL:
+		data_fd = ctx->signal.fid;
+		break;
+	default:
+		return ONI_EPATHINVALID;
+	}
+
+	while (received < size) {
+
+		int rc = read(data_fd, (char *)data + received, size - received);
+
+		if ((rc < 0) && (errno == EINTR))
+			continue;
+
+		if (rc <= 0)
+			return ONI_EREADFAILURE;
+
+		received += rc;
+	}
+
+	return received;
+}
+
+int oni_driver_write_stream(oni_driver_ctx driver_ctx, oni_write_stream_t stream, const char* data, size_t size)
+{
+	CTX_CAST;
+	size_t written = 0;
+
+	int data_fd;
+	switch (stream)
+	{
+	case ONI_WRITE_STREAM_DATA:
+		data_fd = ctx->read.fid;
+		break;
+	default:
+		return ONI_EPATHINVALID;
+	}
+
+	while (written < size) {
+
+		int rc = write(data_fd, data + written, size - written);
+
+		if ((rc < 0) && (errno == EINTR))
+			continue;
+
+		if (rc <= 0)
+			return ONI_EWRITEFAILURE;
+
+		written += rc;
+	}
+
+	return written;
+}
+
+int oni_driver_write_config(oni_driver_ctx driver_ctx, oni_config_t reg, oni_reg_val_t value)
+{
+	CTX_CAST;
+	oni_conf_off_t write_offset = _oni_register_offset(reg);
+	int config_fd = ctx->config.fid;
+
+	if (lseek(config_fd, write_offset, SEEK_SET) < 0)
+		return ONI_ESEEKFAILURE;
+
+	if (write(config_fd, &value, ONI_REGSZ) != ONI_REGSZ)
+		return ONI_EWRITEFAILURE;
+
+	return ONI_ESUCCESS;
+}
+
+int oni_driver_read_config(oni_driver_ctx driver_ctx, oni_config_t reg, oni_reg_val_t* value)
+{
+	CTX_CAST;
+	oni_conf_off_t read_offset = _oni_register_offset(reg);
+	int config_fd = ctx->config.fid;
+
+	if (lseek(config_fd, read_offset, SEEK_SET) < 0)
+		return ONI_ESEEKFAILURE;
+
+	if (read(config_fd, value, ONI_REGSZ) != ONI_REGSZ)
+		return ONI_EREADFAILURE;
+
+	return ONI_ESUCCESS;
+}
+
+//Right now we do not do anything for the common options
+int oni_driver_set_opt_callback(oni_driver_ctx driver_ctx, int oni_option, const void* value, size_t option_len)
```

Comment:
Im not sure I understand the purpose of this function

Reply:
It's called by libONI when the user sets a non-driver context option (it's the last thing `one_set_opt()` does), in case the driver needs to do something. Say post-reset driver-specific operations or changing some kind of internal buffering scheme when setting the global buffer sizes in the main lib.

The xillybus interface is pretty straightforward so it is not needed, but a future driver could.